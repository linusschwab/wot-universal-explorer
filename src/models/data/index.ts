export {Data} from './Data'
export {InputData} from './InputData'
export {OutputData} from './OutputData'
