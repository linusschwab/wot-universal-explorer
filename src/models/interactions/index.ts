export {InteractionData} from './InteractionData';
export {InteractionPattern} from './InteractionPattern';
export {Action} from './Action';
export {Event} from './Event';
export {Property} from './Property';
export {ISubscriber} from './ISubscriber';
