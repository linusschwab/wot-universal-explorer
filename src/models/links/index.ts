export {Link} from './Link';
export {HTTPLink} from './HTTPLink';
export {COAPLink} from './COAPLink';
export {Operation} from './Operation';
