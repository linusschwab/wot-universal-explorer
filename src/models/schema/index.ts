export {DataSchema} from './DataSchema'
export {InputSchema} from './InputSchema'
export {OutputSchema} from './OutputSchema'
