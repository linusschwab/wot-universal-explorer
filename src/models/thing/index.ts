export {Thing} from './Thing';
export {ThingsManager} from './ThingsManager';
