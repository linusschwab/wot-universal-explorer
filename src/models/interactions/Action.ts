import {Interaction} from "./Interaction";
import {InputData, OutputData} from "../data";


export class Action extends Interaction {

    public inputData: InputData;

    constructor(name: string, outputData: OutputData, inputData: InputData) {
        super(name);
        this.outputData = outputData;
        this.inputData = inputData;
    }

    public invoke() {
        // TODO: Implement
        return true;
    }

    public toString() {
        return this.name + ' action';
    }

}
