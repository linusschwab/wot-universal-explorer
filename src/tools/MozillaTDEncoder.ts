import {Thing} from "../models/thing";
import {Action, Event, Property} from "../models/interactions";
import {App} from "../app";


export class MozillaTDEncoder {

    public static encode(thing: Thing) {
        return JSON.stringify(thing, (key: string, value: Object) => {
            if (value instanceof Thing) {
                return MozillaTDEncoder.thing(value);
            }

            return value;
        }, 4);
    }

    private static thing(thing: Thing) {
        // TODO: href? Include in Thing?
        return {
            "name": thing.name,
            "type": thing.type,
            "href": "",
            "properties": MozillaTDEncoder.properties(thing.properties),
            "actions": MozillaTDEncoder.actions(thing.actions),
            "events": MozillaTDEncoder.events(thing.events)
        };
    }

    private static properties(properties: Property[]) {
        let obj: any = {};
        for (const property of properties) {
            obj[property.name] = {
                "type": property.schema.type,
                "description": property.description,
                "href": App.url + '/things' + property.url
            };

            if (!property.description) {
                delete obj.description;
            }
        }
        return obj;
    }

    private static actions(actions: Action[]) {
        let obj: any = {};
        for (const action of actions) {
            // TODO: Include href (as links object?)
            obj[action.name] = {
                "description": action.description,
            };
        }
        return obj;
    }

    private static events(events: Event[]) {
        let obj: any = {};
        for (const event of events) {
            // TODO: Include href (as links object?)
            obj[event.name] = {
                "description": event.description,
            };
        }
        return obj;
    }
}