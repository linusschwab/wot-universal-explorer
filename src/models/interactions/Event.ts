import {InteractionPattern} from "./InteractionPattern";
import {DataSchema} from "../schema";
import {Operation} from "../links";
import {InteractionData} from "./InteractionData";
import {TimeoutError} from "../../tools/errors";


export class Event extends InteractionPattern {

    public schema: DataSchema;
    public data: InteractionData[];

    constructor(name: string, schema: DataSchema) {
        super(name);
        this.schema = schema;
        this.data = [];
    }

    public async update() {
        // Only poll if regular event link
        if (this.links.length === 0) {
            return;
        }

        try {
            // TODO: Choose correct link
            const data = new InteractionData(await this.links[0].execute());
            this.data.push(data);
            this.notifySubscribers(data);
        } catch (e) {
            if (e !instanceof TimeoutError) {
                throw e;
            }
        }
    }

    public getData(newerThan: number = 0) {
        if (!newerThan) {
            return this.data;
        }

        let newData = [];
        for (let data of this.data) {
            if (data.timestamp > newerThan) {
                newData.push(data);
            }
        }
        return newData;
    }

    public toString() {
        return 'Event ' + this.name;
    }

    public get url() {
        return '/' + this.thing.slug + '/events/' + this.slug;
    }

    public get operations(): Operation[] {
        return [new Operation('get', 'List of ' + this.name + ' Event data', this)];
    }
}
