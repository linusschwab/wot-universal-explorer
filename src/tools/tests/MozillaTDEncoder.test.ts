import {HTTPLink} from "../../models/links";
import {DataSchema, InputSchema, OutputSchema} from "../../models/schema";
import {Thing} from "../../models/thing";
import {Action, Property} from "../../models/interactions";
import {MozillaTDEncoder} from "../MozillaTDEncoder";


test('thing with base, 1 property and 1 action', () => {
    // Prepare test thing
    const thing = new Thing('Testthing', 'Thing', 'http://localhost');

    const property = new Property('testProperty', new DataSchema('integer', true), true, false);
    property.description = 'Test Property Description';
    property.registerLink(new HTTPLink('/testproperty', thing.base, 'application/json'));
    thing.registerInteraction(property);

    const action = new Action('testAction', new InputSchema('boolean'), new OutputSchema('boolean'));
    action.description = 'Test Action Description';
    action.registerLink(new HTTPLink('/testaction', thing.base, 'application/json'));
    thing.registerInteraction(action);

    // Encode
    const td = MozillaTDEncoder.encode(thing);
    expect(td).toMatchSnapshot();
});

test('thing without base, 1 property with multiple links', () => {
    // Prepare test thing
    const thing = new Thing('Testthing', 'Thing');

    const property = new Property('testProperty', new DataSchema('integer', true), true, false);
    property.registerLink(new HTTPLink('http://link1.com/testproperty', '', 'application/json'));
    property.registerLink(new HTTPLink('http://link2.com/testproperty', '', 'application/json'));
    thing.registerInteraction(property);

    // Encode
    const td = MozillaTDEncoder.encode(thing);
    expect(td).toMatchSnapshot();
});
