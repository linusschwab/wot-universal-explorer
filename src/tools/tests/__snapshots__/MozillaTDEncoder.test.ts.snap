// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`thing with base, 1 property and 1 action 1`] = `
"{
    \\"name\\": \\"Testthing\\",
    \\"type\\": \\"Thing\\",
    \\"href\\": \\"\\",
    \\"properties\\": {
        \\"testProperty\\": {
            \\"type\\": \\"integer\\",
            \\"description\\": \\"Test Property Description\\",
            \\"href\\": \\"undefined/things/testthing/properties/testproperty\\"
        }
    },
    \\"actions\\": {
        \\"testAction\\": {
            \\"description\\": \\"Test Action Description\\"
        }
    },
    \\"events\\": {}
}"
`;

exports[`thing without base, 1 property with multiple links 1`] = `
"{
    \\"name\\": \\"Testthing\\",
    \\"type\\": \\"Thing\\",
    \\"href\\": \\"\\",
    \\"properties\\": {
        \\"testProperty\\": {
            \\"type\\": \\"integer\\",
            \\"href\\": \\"undefined/things/testthing/properties/testproperty\\"
        }
    },
    \\"actions\\": {},
    \\"events\\": {}
}"
`;
